src/test/storage/encoded_segment_test.cpp

```diff
@@ -303,4 +303,187 @@ TEST_F(EncodedSegmentTest, SegmentReencoding) {
+  EXPECT_EQ(run_length_segment->null_values()->size(), 9 + 10 - 2);
+  EXPECT_EQ(run_length_segment->end_positions()->size(), 9 + 10 - 2);
+  EXPECT_EQ(run_length_segment->end_positions()->front(), 9ul);
+  EXPECT_EQ(run_length_segment->end_positions()->back(), 99ul);
+
+  // Values runs for values 4 and 96 have been basically removed.
+  EXPECT_EQ(run_length_segment->values()->at(3), 3);
+  EXPECT_EQ(run_length_segment->values()->at(4), 5);
+  EXPECT_EQ(run_length_segment->values()->at(13), 95);  // 9 + 5 - 1 (succeeding run of 4 removed from values())
+  EXPECT_EQ(run_length_segment->values()->at(14), 97);  // no value 96 in values()
+
+  // Check that successive NULL runs are merged to single position
+  EXPECT_EQ(run_length_segment->null_values()->at(2), false);
+  EXPECT_EQ(run_length_segment->null_values()->at(3), true);  // NULLs of values 3/4 are merged into single run
+  EXPECT_EQ(run_length_segment->null_values()->at(4), false);
+  EXPECT_EQ(run_length_segment->null_values()->at(12), false);
+  EXPECT_EQ(run_length_segment->null_values()->at(13), true);
+  EXPECT_EQ(run_length_segment->null_values()->at(14), false);
+}
+
+// Testing the internal data structures of Frame of Reference-encoded segments. In particular, the determination of the
+// reference value (i.e., the block minimum) as well as the difference to the reference value are checked. This test
+// does not test the creation of multiple frames as the required vectors are too large for unit testing.
+TEST_F(EncodedSegmentTest, FrameOfReference) {
+  constexpr auto row_count = int32_t{17};
+  constexpr auto minimum = int32_t{5};
+  auto values = pmr_concurrent_vector<int32_t>(row_count);
+  auto null_values = pmr_concurrent_vector<bool>(row_count);
+
+  for (auto row_id = int32_t{0u}; row_id < row_count; ++row_id) {
+    values[row_id] = minimum + row_id;
+  }
+  null_values[1] = true;
+  null_values[7] = true;
+
+  auto values_copy = values;
+  auto null_values_copy = null_values;
+  const auto value_segment = std::make_shared<ValueSegment<int32_t>>(std::move(values), std::move(null_values));
+  const auto encoded_segment =
+      this->encode_segment(value_segment, DataType::Int, SegmentEncodingSpec{EncodingType::FrameOfReference});
+
+  const auto for_segment = std::dynamic_pointer_cast<const FrameOfReferenceSegment<int32_t>>(encoded_segment);
+  ASSERT_TRUE(for_segment);
+  EXPECT_EQ(for_segment->block_minima().size(), 1);  // single block
+  EXPECT_EQ(for_segment->offset_values().size(), row_count);
+  EXPECT_EQ(for_segment->null_values().size(), row_count);
+
+  EXPECT_EQ(for_segment->null_values()[0], false);
+  EXPECT_EQ(for_segment->null_values()[1], true);
+  EXPECT_EQ(for_segment->null_values()[7], true);
+  EXPECT_EQ(for_segment->null_values()[16], false);
+
+  // Block minium should be the smallest value: 0
+  EXPECT_EQ(for_segment->block_minima().front(), minimum);
+  resolve_compressed_vector_type(for_segment->offset_values(), [&](const auto& offset_values) {
```

Comment:
Does this method fail if the vector is not compressed? I.e. do we make sure we use vector compression since FoR encoding would not give use benefits otherwise?

Reply:
Yes, FoR without vector compression would mean that the offsets are uncompressed and that would lead to a netto gain (due to the additional minima vector). So FoR requires vector compression.

Reply:
The call should not fail since `offset_values` is of type `BaseCompressedVector`.